src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/AssetCache.cs

```diff
@@ -178,83 +220,158 @@ public static IEnumerable<T> EnumerateAllComponentsInScenesAndPrefabs<T>()
+#pragma warning restore 414
+
         public TAsset GetAsset(AssetId assetId)
         {
             using (StateSynchronizationPerformanceMonitor.Instance.IncrementEventDuration(this.GetType().Name, "GetAsset"))
             {
-                TAssetEntry assetEntry;
-                if (LookupByAssetId.TryGetValue(assetId, out assetEntry))
+                if (assetId == null ||
+                assetId == AssetId.Empty ||
+                assetId.Name == string.Empty)
+                {
+                    return default(TAsset);
+                }
+
+                if (lookupByAssetId.TryGetValue(assetId, out var entry))
                 {
-                    return assetEntry.Asset;
+                    return (TAsset)entry.Asset;
                 }
-                else
+
+                if (!TryLoadAssets(assetId.Name))
                 {
                     return default(TAsset);
                 }
+
+                if (!lookupByAssetId.TryGetValue(assetId, out entry))
+                {
+                    Debug.LogError($"Assets were loaded for {assetId.Name} but no associated asset cache entry was found for {assetId}");
+                    return default(TAsset);
+                }
+
+                return (TAsset)entry.Asset;
             }
         }
 
         public AssetId GetAssetId(TAsset asset)
         {
             using (StateSynchronizationPerformanceMonitor.Instance.IncrementEventDuration(this.GetType().Name, "GetAssetId"))
             {
-                TAssetEntry assetEntry;
-                if (asset != null && LookupByAsset.TryGetValue(asset, out assetEntry))
+                if (asset == null ||
+                    asset.name == null ||
+                    asset.name == string.Empty)
                 {
-                    return assetEntry.AssetId;
+                    return AssetId.Empty;
                 }
-                else
+
+                if (lookupByAsset.TryGetValue(asset, out var entry))
+                {
+                    return entry.AssetId;
+                }
+
+                if (!TryLoadAssets(asset.name))
                 {
                     return AssetId.Empty;
                 }
+
+                if (!lookupByAsset.TryGetValue(asset, out entry))
+                {
+                    Debug.LogError($"Assets were loaded for {asset.name} but not associated asset cache entry was found for {asset}");
+                    return AssetId.Empty;
+                }
+
+                return entry.AssetId;
+            }
+        }
+
+        public override void LoadAllAssets()
+        {
+            using (StateSynchronizationPerformanceMonitor.Instance.IncrementEventDuration(this.GetType().Name, "LoadAllAssets"))
+            {
+                bool successful = true;
+                foreach (var pair in LookupByName)
+                {
+                    if (!TryLoadAssets(pair.Key))
+                    {
+                        successful = false;

```

Comment:
> successful = false; [](start = 23, length = 20)

Consider logging what the failed asset was (i.e. so instead of seeing "something failed" you'll see "this failed")

Reply:
TryLoadAssets already flags what portion of the look up failed. I don't think we need to duplicate that information here, but we could still keep this asset lookup failure for the overall asset cache.

---
In reply to: [355041870](https://github.com/microsoft/MixedReality-SpectatorView/pull/295#discussion_r355041870) [](ancestors = 355041870)